### PR TITLE
#171380747 User login with email and password

### PR DESCRIPTION
**Pull request by descholar-ceo**
#### What does this PR ​ do​ ?
enables the user to login using email and password

#### Description of Task to be completed?
* -Add findByEmail function to UserService class
* -Add userLogin function to AuthenticationController which holds the real functionality of the login
* -Add Documentations of the API in swagger.json file
* -Write Unit testing for the API in the tests directory
* -Add login router

##### A.Pre-requisites : 
* To be able to test this PR locally, you need to have [PostgreSQL version `9.5.21` or higher](https://www.postgresql.org/) database engine already installed on your local machine.
* You need to have [NodeJS version `6` or higher](https://nodejs.org/en/) and [npm version `6` or higher](https://www.npmjs.com/)
* Your local machine needs to be able to run terminal, powershell, or bash commands

##### B.Steps:
1. If you don't yet have this repository on your local machine, clone it with this command in your terminal `git clone https://github.com/Stackup-Rwanda/stackup2-barefoot-backend`
1. After cloning the repository, run the following command respectively
    > * cd stackup2-barefoot-backend
    > * git fetch origin ft-user-login-email-password-171380747
    > * git checkout ft-user-login-email-password-171380747
    > * npm install
 1. Rename the file `.env.example` to `.env` which is in root directory of this project (current directory if you followed all of the steps respectively)
1.  Give values to these fields 
```javascript
#Development database
DEV_DB_USERNAME=YOUR_USERNAME //username of your PostgreSQL db
DEV_DB_PASSWORD=YOUR_PASSWORD //password you use to login into your PostgreSQL db
DEV_DB_NAME=DATABASE_NAME //name of the database you want to create
DEV_DB_HOSTNAME=YOUR_HOSTNAME //ex: localhost
DEV_DB_PORT=DATABASE_PORT_NUMBER //ex: 5432

#TOKEN TO BE USED WHILE GENERATING JWT_TOKEN
JWT_KEY=your_desired_tokenkey
```
5. After giving values to those fields, then run the following commands in your terminal, respectively:
    > * npx sequelize-cli db:create
    > * npx sequelize-cli db:migrate
    > * npm start    #This command requires that the 3000 port of your local machine must be free
1. If you have [Postman](https://www.postman.com/) installed on your local machine, then send a POST request to this URL = `localhost:3000/api/auth/signup` with Content-Type in headers `Application/Json` with data in the requestBody : 
```javascript
{
  "firstName": "John",
  "lastName": "Doe",
  "email": "john@doe.com",
  "address": "Kigali, Rwanda",
  "password": "hellothere",
  "gender": "Male"
}
```
=>All of the above steps leads you to the successful Signup

=> So, to login, then send a POST request to this URL = `localhost:3000/api/auth/login` with Content-Type in headers `Application/Json` with data in the requestBody : 
```javascript
{
  "email": "john@doe.com",
  "password": "hellothere",
}
```

=>If you enter wrong data, or empty data, the screenshot section of this PR shows you screenshots of different responses you will get.
#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#171380747](https://www.pivotaltracker.com/n/projects/2435482)
##### Successful login
Screenshot from 2020-03-05 21-42-30

##### Login with wrong email or password
Screenshot from 2020-03-05 21-44-17

##### Login with empty data
Screenshot from 2020-03-05 21-46-02

#### Questions:
N/A

### CODE REVIEW

**Review by since9629 — COMMENTED** [review]
@de-scholar Could you add how this PR should be tested

**Comment by descholar-ceo** [comment]
@karamuka I added how you should test it manualy

**Review by Muhire-Josue — COMMENTED** [review]
Can you change the `what this PR do?` response? your response was  `deploy to heroku` I think it's `enables the user to login using email and password`

File: src/controllers/login.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
+import UserService from '../services/UserService';
+import { isPasswordTrue } from '../utils/paaswordDecrypt';
```
Comment (Muhire-Josue) [inline src/controllers/login.js]: can you rename this file `../utils/paaswordDecrypt` to `../utils/passwordDecrypt`. It's just a little misspell there :)

File: src/controllers/login.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import generateToken from '../utils/tokenHandler';
+
+/**
+   * function findAll() returns all users in db
+   * @param {object} req it request
+   * @param {object} res it response
+   * @returns {object} send reponse to the user
+   */
+const getUser = async (req, res) => {
```
Comment (Muhire-Josue) [inline src/controllers/login.js]: you might also want to rename this arrow function to something more descriptive like `userLogin`

File: src/controllers/login.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+          names: dataValues.names,
+          email: dataValues.email
+        };
+        const token = await generateToken(userInfo);
+        res.status(200).json({
+          message: 'successfully logged in',
+          token
+        });
+      } else {
+        res.status(401).json({
```
Comment (Muhire-Josue) [inline src/controllers/login.js]: Could you create a middleware to check both email and password you can call it `credentialChecker` and call it in the route like this `router.post('/api/user/', crendentialChecker, login.userLogin)`, and move all the 401 from controller to that middleware.

File: src/controllers/login.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    if (gottenUser) {
+      const { dataValues } = gottenUser;
+      if (await isPasswordTrue(password, dataValues.password)) {
+        const userInfo = {
+          names: dataValues.names,
+          email: dataValues.email
+        };
+        const token = await generateToken(userInfo);
+        res.status(200).json({
```
Comment (Muhire-Josue) [inline src/controllers/login.js]: could you create a helper function to manage the responses? you can refer to @the22mastermind's PR

File: src/controllers/login.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+          message: 'unknown credentials'
+        });
+      }
+    } else {
+      res.status(401).json({
+        message: 'unknown credentials'
+      });
+    }
+  } catch (err) {
+    res.status(400).json({ message: 'enter your email or password' });
```
Comment (Muhire-Josue) [inline src/controllers/login.js]: You could also create a middleware for this.

File: src/controllers/login.js (hunk 1 above)
Comment (descholar-ceo) [reply src/controllers/login.js]: done

File: src/controllers/login.js (hunk 4 above)
Comment (descholar-ceo) [reply src/controllers/login.js]: done, and thank you so much, this helped my function to be smaller than before

File: src/controllers/login.js (hunk 1 above)
Comment (since9629) [reply src/controllers/login.js]: Great @descholar-ceo, would you push your changes then

File: src/controllers/login.js (hunk 2 above)
Comment (descholar-ceo) [reply src/controllers/login.js]: done

File: src/controllers/login.js (hunk 3 above)
Comment (descholar-ceo) [reply src/controllers/login.js]: I refactored my codes to make the function easy but didn't create the middleware, you can check the way I did, and you tell me how you think about it.

**Review by Muhire-Josue — APPROVED** [review]
LGTM

File: src/controllers/login.js (hunk 3 above)
Comment (descholar-ceo) [reply src/controllers/login.js]: done

**Review by muchristian — APPROVED** [review]

**Review by ricahcyuzuzo — APPROVED** [review]
Good Job!

**Review by since9629 — APPROVED** [review]

**Review by muchristian — CHANGES_REQUESTED** [review]
i would suggest you to replace verifyPassword and verifyEmail function from `middleware->credentialChecker` to `controller->authenticationController->userLogin function`. read the following paragraph you will understand the reason why i told you to replace them as aforementioned **:**`middleware are often reused more than once and often they do not response. On contrary, controller respond and are most of the time specific to one endpoint.`

File: src/controllers/authentication.controller.js [hunk 6]
```diff
@@ -37,4 +37,16 @@ export default class AuthenticationControler {
       return errorResponse(res, statusCodes.badRequest, customMessages.userSignupFailed);
     }
   }
+
+  /**
+   * function findAll() returns all users in db
```
Comment (mifeille) [inline src/controllers/authentication.controller.js]: does really this function find all users in the db?

Comment (descholar-ceo) [reply src/controllers/authentication.controller.js]: @mifeille I changed the description, it was an error I did

**Comment by descholar-ceo** [comment]
@muchristian I refactored my codes now you can re-check

**Review by muchristian — APPROVED** [review]

**Pull request merged**